### PR TITLE
feat: add text & numeric filters to insights routing responses

### PR DESCRIPTION
**Pull request by eunjae-lee**
## What does this PR do?

- Fixes #XXXX (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A - I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

**Comment by sean-brydon** [comment]
I've tested a lot of this. From text filters to number filters to old filters. All seems to be working as expected. 
CleanShot 2024-12-11 at 08 31 15
Lets disable / fix filtering on the assignment reason (this is just a text string that gets returned)

File: apps/web/public/static/locales/en/common.json [hunk 1]
```diff
@@ -2838,6 +2838,12 @@
   "filter_operator_ends_with": "Ends with",
   "filter_operator_is_empty": "Is empty",
   "filter_operator_not_empty": "Not empty",
+  "filter_operator_eq": "=",
+  "filter_operator_neq": "≠",
+  "filter_operator_gt": ">",
+  "filter_operator_gte": "≥",
+  "filter_operator_lt": "<",
+  "filter_operator_lte": "≤",
```
Comment (sean-brydon) [inline apps/web/public/static/locales/en/common.json:2846]: NIT: IM not sure these need to be translated. I don't think anywhere in the world uses different symbols to this

**Review by sean-brydon — COMMENTED** [review]
Slight nit: On routing/insights when de-selecting a user the people filter gets cleared. This is not the case on normal insights

Also i found during testing this: we should send "asc"/"desc" from the sorting state of the table to the backend filters here too. You can sort say ratings by "asc" client side and when loading in data as you scroll the table jumps around a bit as it moves the new loaded data into the table state (Do not fix in this PR - we can follow up)

File: apps/web/public/static/locales/en/common.json (hunk 1 above)
Comment (eunjae-lee) [reply apps/web/public/static/locales/en/common.json:2846]: ah that's a good point!

**Comment by eunjae-lee** [comment]
> On routing/insights when de-selecting a user the people filter gets cleared

1.
@sean-brydon  Actually I thought about this. The `FilterSelect` component doesn't differentiate unchecking an item vs clicking "Clear filter" button. Maybe I could add an extra `onCleared()` prop to the component, but it's being used in many places. So I didn't do it in this PR. (Maybe when we re-work all the filters)

Screenshot 2024-12-11 at 11 28 10

2.
Also, about the server-side ordering, yes, it's on my list. I'll tackle it separately!

File: packages/features/data-table/lib/types.ts [hunk 2]
```diff
@@ -1,5 +1,7 @@
 import { z } from "zod";
 
+import type { IconName } from "@calcom/ui/components/icon/icon-names";
```
Comment (emrysal) [inline packages/features/data-table/lib/types.ts:3]:
```suggestion
import type { IconName } from "@calcom/ui/icon";
```

File: packages/features/data-table/lib/utils.ts [hunk 3]
```diff
@@ -1,51 +1,77 @@
...
 
-export function useFiltersSearchState() {
-  return useQueryStates(filtersSearchParams);
+export function useFilterValue<T>(columnId: string, schema: z.ZodType<T>) {
+  const { activeFilters } = useDataTable();
+  return useMemo(() => {
+    const value = activeFilters.find((filter) => filter.f === columnId)?.v;
+    if (schema && value) {
+      try {
+        return schema.parse(value);
```
Comment (emrysal) [inline packages/features/data-table/lib/utils.ts:30]:
```suggestion
        return schema.safeParse(value);
```

File: packages/features/insights/components/RoutingFormResponsesTable.tsx [hunk 4]
```diff
@@ -41,6 +63,17 @@ type RoutingFormTableRow = {
...
+type FieldCellValue = { label: string; value: string };
+
+const ZResponseValues = z.array(
+  z.object({
+    label: z.string(),
+    value: z.string(),
+  })
+);
+
+type ResponseValues = z.infer<typeof ZResponseValues>;
```
Comment (emrysal) [inline packages/features/insights/components/RoutingFormResponsesTable.tsx:75]: If not needed, don't rely on z.infer.

```suggestion
type ResponseValues = FieldCellValue[];
```

File: packages/features/insights/server/trpc-router.ts [hunk 5]
```diff
@@ -1651,12 +1645,16 @@ export const insightsRouter = router({
       })
     )
     .query(async ({ ctx, input }) => {
-      return await RoutingEventsInsights.getRoutingFormHeaders({
+      const fields = await RoutingEventsInsights.getRoutingFormHeaders({
```
Comment (emrysal) [inline packages/features/insights/server/trpc-router.ts:1648]: Naming wise this causes me some confusion:

```suggestion
      const headers = await RoutingEventsInsights.getRoutingFormHeaders({
```

**Review by sean-brydon — COMMENTED** [review]
LGTM - comment instead of approval so it doesnt get accidentally merged

**Review by emrysal — DISMISSED** [review]
it's time 🕐

**Review by emrysal — APPROVED** [review]

**Pull request merged**